<?xml version="1.0"?>
<!-- 
This launch file initializes the scan_and_approach_node which combines
the functionality of move_to_pose_scan and approach_object nodes.

Command to use with interbotix nav stack:
roslaunch interbotix_xslocobot_nav xslocobot_nav_2.launch robot_model:=locobot_wx250s use_lidar:=true localization:=true
-->
<launch>
  <!-- Global Arguments -->
  <arg name="robot_name" default="locobot"/>
  <arg name="rviz_frame" default="map"/>
  <arg name="rvizconfig" default="$(find yolo_vision)/rviz/xslocobot_description_yolo.rviz"/>
  <arg name="use_rviz" default="true" doc="Set to false to disable RViz visualization"/>
  <arg name="debug_mode" default="false"/>

  <!-- Camera Topics -->
  <arg name="rgb_topic" default="/$(arg robot_name)/camera/color/image_raw"/>
  <arg name="depth_topic" default="/$(arg robot_name)/camera/depth/image_rect_raw"/>


  <!-- YOLO Topics -->
  <arg name="model_path" default="$(find yolo_vision)/models/simple_bbox_detection.pt"/>
  <arg name="annotated_image_topic" default="camera/yolo/annotated_image"/>
  <arg name="bbox_depth_topic" default="camera/yolo/bbox_depth"/>
  <arg name="camera_info_topic" default="camera/color/camera_info"/>
  <arg name="object_marker_topic" default="object_markers"/>
  <arg name="keypoints_topic" default="camera/yolo/keypoints"/>
  <arg name="keypoints_viz_topic" default="camera/yolo/keypoints_visualization"/>
  <arg name="confidence_threshold" default="0.5"/>
  <arg name="input_size" default="640"/>
  <arg name="camera_frame" default="$(arg robot_name)/camera_color_optical_frame"/>
  <arg name="keypoint_model_path" default="$(find yolo_vision)/models/grasshopper_and_boxes_keypoint.pt"/>

  <!-- Move and scan parameters -->
  <arg name="poses_config" default="$(find move_to_pose)/config/poses.yaml"/>
  <arg name="pose_command" default="all"/>
  <arg name="move_base_topic" default="/$(arg robot_name)/move_base"/>
  <arg name="costmap_topic" default="/$(arg robot_name)/move_base/global_costmap/costmap"/>

  <!-- Launch YOLO node -->
  <node name="yolo_node" pkg="yolo_vision" type="yolo_node.py" output="screen" ns="$(arg robot_name)">

      <!-- input params -->
      <param name="debug_mode" value="$(arg debug_mode)"/>

      <!-- Input Topics -->
      <param name="image_topic" value="$(arg rgb_topic)"/>
      <param name="model_path" value="$(arg model_path)"/>
      <param name="depth_image_topic" value="$(arg depth_topic)"/>

      <!-- Output Topics -->
      <param name="annotated_image_topic" value="$(arg annotated_image_topic)"/>
      <param name="bbox_depth_topic" value="$(arg bbox_depth_topic)"/>

      <!-- Rviz -->
      <param name="use_rviz" value="flase"/>
  </node>

  <!-- Object Mapper Node -->
  <node name="object_mapper" pkg="yolo_vision" type="object_mapper_node.py" output="screen" ns="$(arg robot_name)">

    <!-- Input Topics -->
    <param name="camera_frame" value="$(arg robot_name)/camera_color_optical_frame"/>
    <param name="map_frame" value="map"/>
    <param name="bbox_depth_topic" value="$(arg bbox_depth_topic)"/>
    <param name="camera_info_topic" value="$(arg camera_info_topic)"/>

    <!-- Output Topic -->
    <param name="object_marker_topic" value="$(arg object_marker_topic)"/>
  </node>

    <!-- PART 4: Launch Keypoint Activation Service -->
  <node name="keypoint_activation_node" pkg="yolo_vision" type="yolo_keypoint_activation.py" output="screen" ns="$(arg robot_name)">
      <param name="robot_name" value="$(arg robot_name)"/>
  </node>

  <!-- PART 5: Launch Keypoint Detection node (initially disabled) -->
  <node name="keypoint_detection_node" pkg="yolo_vision" type="yolo_keypoint_detection_node.py" output="screen" ns="$(arg robot_name)">
      <!-- Main parameters -->
      <param name="model" value="$(arg keypoint_model_path)"/>
      <param name="image_topic" value="$(arg rgb_topic)"/>
      <param name="confidence_threshold" value="$(arg confidence_threshold)"/>
      <param name="input_size" value="$(arg input_size)"/>
      <param name="debug_mode" value="$(arg debug_mode)"/>
      
      <!-- Output topics -->
      <param name="keypoints_topic" value="$(arg keypoints_topic)"/>
      <param name="keypoints_viz_topic" value="$(arg keypoints_viz_topic)"/>
      
      <!-- Camera frame -->
      <param name="camera_frame" value="$(arg camera_frame)"/>
  </node>

  <!-- RViz (conditionally launched) -->
  <group if="$(arg use_rviz)">
      <node name="rviz" pkg="rviz" type="rviz" 
            args="-f $(arg rviz_frame) -d $(arg rvizconfig)"
            ns="$(arg robot_name)">
          <remap from="/clicked_point" to="clicked_point"/>
          <remap from="/initialpose" to="initialpose"/>
          <remap from="/move_base_simple/goal" to="move_base_simple/goal"/>
      </node>
  </group>
</launch>